File: HappyPuppy/C#/UseCases.cs
using System.Collections.Generic;


public abstract class UseCase
{
    protected HappyPuppyRepository repo; 

    public UseCase(HappyPuppyRepository repo)
    {
       this.repo = repo;
    }
}

public abstract class DisplayListUseCase: UseCase, ActionRequest
{
    public abstract string Heading { get; }

    public DisplayListUseCase(HappyPuppyRepository repo): base(repo)
    {
    }

    public void Execute(DisplayResponse port) 
    {
       foreach(string s in GetList())
       {
            port(s);
       }
    }

    protected abstract IEnumerable<string> GetList();
}

public class ListPuppiesUseCase: DisplayListUseCase
{ 
    public override string Heading
    {
        get { return "AVAILABLE PUPPIES"; }
    }

    public ListPuppiesUseCase(HappyPuppyRepository repo): base(repo)
    {
    }

    protected override IEnumerable<string> GetList()
    {
        return repo.RetrievePuppies();
    }
}

public class ListOwnersUseCase: DisplayListUseCase
{  
    public override string Heading
    {
        get { return "KNOWN OWNERS"; }
    }

    public ListOwnersUseCase(HappyPuppyRepository repo): base(repo)
    {
    }

    protected override IEnumerable<string> GetList()
    {
        return repo.RetrieveOwners();
    }
}

public class AdoptPuppyUseCase: UseCase, AdoptPuppyRequest
{    
    public string Heading
    {
        get { return "ADOPTION IN PROGRESS"; }
    }

    public AdoptPuppyUseCase(HappyPuppyRepository repo): base(repo)
    {
    }

    public void Execute(string puppyName, string ownerName,
                        NotificationResponse outPort)
    {
            repo.RemovePuppy(puppyName);
            repo.AddOwner(ownerName);
            outPort();
    }
}

File: HappyPuppy/Java/UseCases.java

public class UseCases
{
    public static abstract class UseCase
    {
        protected Repository.HappyPuppyRepository repo; 
    
        public UseCase(Repository.HappyPuppyRepository repo)
        {
           this.repo = repo;
        }
    }

    public static abstract class DisplayListUseCase 
        extends UseCase 
        implements Boundaries.ActionRequest
    {
        public abstract String Heading();
    
        public DisplayListUseCase(Repository.HappyPuppyRepository repo)
        {
            super(repo);
        }
    
        public void Execute(Boundaries.DisplayResponse port) 
        {
           for(String s: GetList())
           {
                port.Accept(s);
           }
        }
    
        protected abstract Iterable<String> GetList();
    }

    public static class ListPuppiesUseCase extends DisplayListUseCase
    { 
        @Override
        public String Heading()
        {
            return "AVAILABLE PUPPIES";
        }
    
        public ListPuppiesUseCase(Repository.HappyPuppyRepository repo)
        {
            super(repo);
        }

        @Override    
        protected Iterable<String> GetList()
        {
            return repo.RetrievePuppies();
        }
    }

    public static class ListOwnersUseCase extends DisplayListUseCase
    {  
        @Override
        public String Heading()
        {
           return "KNOWN OWNERS"; 
        }
    
        public ListOwnersUseCase(Repository.HappyPuppyRepository repo)
        {
            super(repo);
        }
    
        @Override
        protected Iterable<String> GetList()
        {
            return repo.RetrieveOwners();
        }
    }

    public static class AdoptPuppyUseCase 
        extends UseCase 
        implements Boundaries.AdoptPuppyRequest
    {    
        public String Heading()
        {
           return "ADOPTION IN PROGRESS"; 
        }
    
        public AdoptPuppyUseCase(Repository.HappyPuppyRepository repo)
        {    
            super(repo);
        }
    
        public void Execute(String puppyName, String ownerName,
                            Boundaries.NotificationResponse outPort)
        {
                repo.RemovePuppy(puppyName);
                repo.AddOwner(ownerName);
                outPort.ActionWasCompleted();
        }
    }

}
